File: app/src/main/java/com/ps/simplepapersoccer/GameObjects/Game/GameBoard.java
package com.ps.simplepapersoccer.GameObjects.Game;

import com.google.common.collect.Iterables;
import com.ps.simplepapersoccer.Enums.NodeTypeEnum;
import com.ps.simplepapersoccer.GameObjects.Move.PartialMove;
import com.ps.simplepapersoccer.GameObjects.Move.PossibleMove;
import com.ps.simplepapersoccer.Helpers.MathHelper;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.UUID;

public class GameBoard
{
    public HashMap<UUID, Node> nodeHashMap = new HashMap<>();

    public Node ballNode;

    public Node goalNode1;
    public Node goalNode2;

    private ArrayList<PartialMove> allPartialMoves = new ArrayList<>();

    @Override
    public int hashCode() {
        return allPartialMoves.hashCode() ^ nodeHashMap.hashCode() ^ ballNode.hashCode();
    }

    public GameBoard (int gridSizeX, int gridSizeY)
    {
        makeNodes(gridSizeX, gridSizeY);
        ballNode = findNodeByXY(gridSizeX/2, gridSizeY/2);
    }

    private void makeNodes(int gridSizeX, int gridSizeY)
    {
        for (int y = 1; y < gridSizeY; ++y)
        {
            for (int x = 0; x <= gridSizeX; ++x)
            {
                //No node in the 4 cornes
                if (x == 0 && y == 1) continue;
                if (x == 0 && y == gridSizeY - 1) continue;
                if (x == gridSizeX && y == 1) continue;
                if (x == gridSizeX && y == gridSizeY - 1) continue;

                //Check if wall
                if (x == 0 || x == gridSizeX )
                    addNodeToNodeMap(new Node(x,y, NodeTypeEnum.Wall));

                    //Check if top wall special case
                else if (x != gridSizeX / 2 && y == 1)
                {
                    addNodeToNodeMap(new Node(x,y, NodeTypeEnum.Wall));
                }

                //Check if bottom wall special case
                else if (x != gridSizeX / 2 && y == gridSizeY - 1)
                {
                    addNodeToNodeMap(new Node(x,y, NodeTypeEnum.Wall));
                }

                //Regular empty node
                else
                {
                    addNodeToNodeMap(new Node(x,y, NodeTypeEnum.Empty));
                }
            }
        }
        //Make the 2 goal nodes
        goalNode1 = new Node(gridSizeX / 2, gridSizeY, NodeTypeEnum.Goal);
        addNodeToNodeMap(goalNode1);

        goalNode2 = new Node(gridSizeX / 2, 0, NodeTypeEnum.Goal);
        addNodeToNodeMap(goalNode2);


        GenerateAllNeighbors();
    }

    public void GenerateAllNeighbors()
    {
        for (Node node : nodeHashMap.values()) {
            for (Node otherNode : nodeHashMap.values()) {
                if (node.id == otherNode.id) continue;

                double euclideanDistance = MathHelper.euclideanDistance(node.xCord, otherNode.xCord, node.yCord, otherNode.yCord);

                if (node.nodeType == NodeTypeEnum.Wall && otherNode.nodeType == NodeTypeEnum.Wall) {
                    if (node.yCord != otherNode.yCord && otherNode.xCord != node.xCord && euclideanDistance < 2) {
                        node.AddNeighborPair(otherNode);
                    } else {
                        continue;
                    }
                }
                if (euclideanDistance < 2) node.AddNeighborPair(otherNode);
            }
        }
    }

    public HashSet<PossibleMove> allPossibleMovesFromNode(Node node)
    {
        HashSet<PossibleMove> possibleMoves = new HashSet<>();

        for (UUID uuid : node.neighbors) {
            Node otherNode = nodeHashMap.get(uuid);
            possibleMoves.add(new PossibleMove(node, otherNode));
        }

        return possibleMoves;
    }

    public PartialMove UndoLastMove()
    {
        PartialMove partialMove = Iterables.getLast(allPartialMoves);

        partialMove.newNode.AddNeighborPair(partialMove.oldNode);

        nodeHashMap.get(partialMove.newNode.id).nodeType = partialMove.newNode.nodeType;

        ballNode = partialMove.oldNode;

        allPartialMoves.remove(partialMove);

        return partialMove;
    }

    public void MakePartialMove(PartialMove partialMove)
    {
        allPartialMoves.add(new PartialMove(new Node(partialMove.oldNode), new Node(partialMove.newNode), partialMove.madeTheMove));

        partialMove.newNode.RemoveNeighborPair(partialMove.oldNode);

        if (partialMove.oldNode.nodeType == NodeTypeEnum.Empty)
            partialMove.oldNode.nodeType = NodeTypeEnum.BounceAble;

        ballNode = partialMove.newNode;
    }

    //Returns the node with the XY coordinates
    public Node findNodeByXY(int x, int y)
    {
        for (Node n : nodeHashMap.values())
        {
            if (n.xCord == x && n.yCord == y)
                return n;
        }
        return null;
    }

    public void addNodeToNodeMap(Node node)
    {
        nodeHashMap.put(node.id, node);
    }
}

File: app/src/main/java/com/ps/simplepapersoccer/GameObjects/Game/GameBoard.kt
package com.ps.simplepapersoccer.gameObjects.game

import com.ps.simplepapersoccer.data.enums.ConnectionTypeEnum
import com.ps.simplepapersoccer.data.enums.NodeTypeEnum
import com.ps.simplepapersoccer.gameObjects.game.geometry.ConnectionNode
import com.ps.simplepapersoccer.gameObjects.game.geometry.Goal
import com.ps.simplepapersoccer.gameObjects.game.geometry.Node
import com.ps.simplepapersoccer.gameObjects.game.geometry.TwoDimensionalPoint
import com.ps.simplepapersoccer.gameObjects.game.geometry.abstraction.BaseNode
import com.ps.simplepapersoccer.gameObjects.game.geometry.abstraction.IntegerLine
import com.ps.simplepapersoccer.gameObjects.move.PartialMove
import com.ps.simplepapersoccer.gameObjects.move.PossibleMove
import com.ps.simplepapersoccer.gameObjects.move.StoredMove
import com.ps.simplepapersoccer.helpers.MathHelper
import java.util.*
import kotlin.collections.HashSet
import kotlin.math.roundToInt

data class GameBoard(val gridSizeX: Int, val gridSizeY: Int) {
    val allBaseNodes = mutableListOf<BaseNode>()
    val allNodes: HashSet<Node>

    private var currentBallNode: Node? = null
        set(value) {
            field = value
            allLegalMovesFromBallNode = ballNode.connectedNodes.sortedBy { it.coords }.map { PossibleMove(ballNode, it) }
        }

    val ballNode: Node get() = currentBallNode!!

    var allLegalMovesFromBallNode: List<PossibleMove> = emptyList(); private set

    var currentPlayersTurn: Int = 1

    lateinit var goal1: Goal
    lateinit var goal2: Goal
    private val goalScalingX = (gridSizeX / 6).toDouble().roundToInt()

    var topGoalLines: Goal = Goal(IntegerLine(TwoDimensionalPoint(gridSizeX / 2 - goalScalingX, 0),
            TwoDimensionalPoint(gridSizeX / 2 + goalScalingX, 0)),
            IntegerLine(TwoDimensionalPoint(gridSizeX / 2 - goalScalingX, 0), TwoDimensionalPoint(gridSizeX / 2 - goalScalingX, 2)),
            IntegerLine(TwoDimensionalPoint(gridSizeX / 2 + goalScalingX, 0), TwoDimensionalPoint(gridSizeX / 2 + goalScalingX, 2)),
            IntegerLine(TwoDimensionalPoint(gridSizeX / 2 - goalScalingX, 1), TwoDimensionalPoint(gridSizeX / 2 + goalScalingX, 1)))
    var bottomGoalLines: Goal = Goal(IntegerLine(TwoDimensionalPoint(gridSizeX / 2 - goalScalingX, gridSizeY + 2),
            TwoDimensionalPoint(gridSizeX / 2 + goalScalingX, gridSizeY + 2)),
            IntegerLine(TwoDimensionalPoint(gridSizeX / 2 - goalScalingX, gridSizeY + 2), TwoDimensionalPoint(gridSizeX / 2 - goalScalingX, gridSizeY)),
            IntegerLine(TwoDimensionalPoint(gridSizeX / 2 + goalScalingX, gridSizeY + 2), TwoDimensionalPoint(gridSizeX / 2 + goalScalingX, gridSizeY)),
            IntegerLine(TwoDimensionalPoint(gridSizeX / 2 - goalScalingX, gridSizeY + 1), TwoDimensionalPoint(gridSizeX / 2 + goalScalingX, gridSizeY + 1)))

    var allPartialMoves = Stack<StoredMove>(); private set

    val isGameOver: Boolean
        get() {
            return getWinner(ballNode)
        }

    init {
        makeNodes(gridSizeX, gridSizeY)
        allNodes = allBaseNodes.filterIsInstance<Node>().toHashSet()

        val startingBallNode = allNodes.toList().sortedBy { it.coords }[allNodes.size / 2]
        startingBallNode.containsBall = true
        currentBallNode = startingBallNode
    }

    private fun isEdgeNode(point: TwoDimensionalPoint): Boolean {
        return (point.x == 0 || point.x == gridSizeX || point.y == 2 || point.y == gridSizeY)
    }

    private fun makeNodes(gridSizeX: Int, gridSizeY: Int) {
        for (y in 2..gridSizeY) {
            (0..gridSizeX)
                    .filter { !topGoalLines.contains(TwoDimensionalPoint(it, y)) && !bottomGoalLines.contains(TwoDimensionalPoint(it, y)) }
                    .forEach {
                        when {
                            it % 2 == 1 -> {
                                allBaseNodes.add(ConnectionNode(TwoDimensionalPoint(it, y)))
                            }
                            y % 2 == 1 -> {
                                allBaseNodes.add(ConnectionNode(TwoDimensionalPoint(it, y)))
                            }
                            it % 2 != 1 && y % 2 != 1 -> {
                                if (isEdgeNode(TwoDimensionalPoint(it, y))) {
                                    allBaseNodes.add(Node(TwoDimensionalPoint(it, y), NodeTypeEnum.Wall))
                                } else allBaseNodes.add(Node(TwoDimensionalPoint(it, y), NodeTypeEnum.Empty))
                            }
                        }
                    }
        }

        topGoalLines.allNodes.forEach {
            allBaseNodes.add(it)
        }

        bottomGoalLines.allNodes.forEach {
            allBaseNodes.add(it)
        }

        goal1 = bottomGoalLines
        goal2 = topGoalLines

        generateAllNeighbors()
        allBaseNodes.sortBy { it.coords }
    }

    private fun generateAllNeighbors() {
        for (node in allBaseNodes) {
            for (otherNode in allBaseNodes) {
                if (node == otherNode) continue

                if (node is Node && otherNode is ConnectionNode) {
                    val euclideanDistance = MathHelper.euclideanDistance(node.coords, otherNode.coords)
                    if (euclideanDistance.toInt() < 2) {
                        node.addNeighbor(otherNode)
                        otherNode.neighbors.add(node)
                    }
                }
            }
        }

        for (node in allBaseNodes) {
            if (node is ConnectionNode) {
                node.createNodeConnections()
            }
        }
    }

    private fun getWinner(node: Node): Boolean {
        return when {
            node.nodeType == NodeTypeEnum.Goal -> true
            node.connectedNodes.size == 0 -> {
                true
            }
            else -> {
                false
            }
        }
    }

    private fun allPossibleMovesFromNode(node: Node): List<PossibleMove> {
        return node.connectedNodes.sortedBy { it.coords }.map { PossibleMove(node, it) }
    }

    fun allPossibleMovesFromNodeCoords(node: Node): List<Pair<PossibleMove, Boolean>> {
        val legalMoves = allPossibleMovesFromNode(node)

        return node.coordNeighbors.sortedBy { it.coords }.map { neighborNode ->
            Pair(PossibleMove(node, neighborNode), legalMoves.any { it.newNode == neighborNode })
        }
    }

    fun undoLastMove(): PartialMove? {
        return if (allPartialMoves.isNotEmpty()) {
            val storedMove = allPartialMoves.pop()
            getConnectionNodeBetweenNodes(storedMove.partialMove.oldNode, storedMove.partialMove.newNode)
                    ?.connectNodes(storedMove.partialMove.oldNode, storedMove.partialMove.newNode)

            storedMove.partialMove.newNode.nodeType = storedMove.newNodeTypeEnum

            storedMove.partialMove.oldNode.containsBall = true
            storedMove.partialMove.newNode.containsBall = false
            currentBallNode = storedMove.partialMove.oldNode

            storedMove.partialMove

            currentPlayersTurn = storedMove.partialMove.madeTheMove

            storedMove.partialMove
        } else null
    }

    fun makePartialMove(partialMove: PartialMove) {
        allPartialMoves.add(StoredMove(partialMove, partialMove.oldNode.nodeType, partialMove.newNode.nodeType))

        getConnectionNodeBetweenNodes(partialMove.oldNode, partialMove.newNode)?.disconnectNode(partialMove.oldNode, partialMove.newNode)

        if (partialMove.oldNode.nodeType == NodeTypeEnum.Empty)
            partialMove.oldNode.nodeType = NodeTypeEnum.BounceAble

        partialMove.newNode.containsBall = true
        partialMove.oldNode.containsBall = false
        currentBallNode = partialMove.newNode

        if (partialMove.newNode.nodeType == NodeTypeEnum.Empty) {
            changeTurn()
        }
    }

    private fun changeTurn() {
        currentPlayersTurn = 3 - currentPlayersTurn
    }

    fun findNodeByCoords(point: TwoDimensionalPoint): Node? {
        return allNodes.firstOrNull { it.coords == point }
    }

    private fun getConnectionNodeBetweenNodes(node: Node, node2: Node): ConnectionNode? {
        return node.neighbors.first { it.getNodeConnection(node, node2) != null }
    }

    override fun toString(): String {
        var returnString = ""
        var currentY = 0

        for (i in allBaseNodes.indices) {
            val node = allBaseNodes[i]

            if (currentY != node.coords.y) {
                returnString += "\n"
                ++currentY
            }

            if (node is Node) {
                val nodeTypeSymbol = when (node.nodeType) {
                    NodeTypeEnum.ContainsBall -> "O"
                    NodeTypeEnum.Goal -> "G"
                    NodeTypeEnum.Empty -> "X"
                    NodeTypeEnum.Wall -> "|"
                    NodeTypeEnum.BounceAble -> "B"
                    else -> "?"
                }

                returnString = "$returnString $nodeTypeSymbol"
            } else if (node is ConnectionNode) {
                val nodeTypeSymbol = when (node.connectionTypeEnum) {
                    ConnectionTypeEnum.Open -> "-"
                    ConnectionTypeEnum.Blocked -> "!"
                    ConnectionTypeEnum.LineTiltedLeft -> "\\"
                    ConnectionTypeEnum.LineTiltedRight -> "/"
                    else -> "?"
                }

                returnString = "$returnString $nodeTypeSymbol"
            }

            if (node.coords.x + node.coords.y != i) returnString += " "
        }

        returnString += "\n"

        return returnString
    }
}
